drv/lpc55-rng/src/main.rs

```diff
@@ -2,81 +2,146 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
-//! Random Number Generation
+//! Driver for the LPC55 random number generator.
 //!
-//! This task will produce random u32 values for you, if you ask nicely.
-//!
-//! An example:
-//!
-//! ```ignore
-//! #[derive(AsBytes)]
-//! #[repr(C)]
-//! struct FetchRandomNumber;
-//!
-//! impl hl::Call for FetchRandomNumber {
-//!     const OP: u16 = 0;
-//!     type Response = u32;
-//!     type Err = u32;
-//! }
-//!
-//! let num = hl::send(rng, &FetchRandomNumber).expect("could not ask the rng for a number");
-//!
-//! hprintln!("got {} from the rng", num).ok();
-//! ```
+//! Use the rng-api crate to interact with this driver.
 
 #![no_std]
 #![no_main]
 
+use core::mem::size_of;
 use drv_lpc55_syscon_api::{Peripheral, Syscon};
+use drv_rng_api::RngError;
+use idol_runtime::{ClientError, RequestError};
+use rand_core::block::{BlockRng, BlockRngCore};
+use rand_core::{Error, RngCore};
 use userlib::*;
-use zerocopy::AsBytes;
 
 use lpc55_pac as device;
 
 task_slot!(SYSCON, syscon_driver);
 
-#[repr(u32)]
-enum ResponseCode {
-    BadArg = 1,
-    PoweredOff = 2,
+struct Lpc55Core {
+    pmc: &'static lpc55_pac::pmc::RegisterBlock,
+    rng: &'static lpc55_pac::rng::RegisterBlock,
+    syscon: Syscon,
 }
 
-impl From<ResponseCode> for u32 {
-    fn from(rc: ResponseCode) -> Self {
-        rc as u32
+impl Lpc55Core {
+    fn new() -> Self {
+        let syscon = SYSCON.get_task_id();
+        Lpc55Core {
+            pmc: unsafe { &*device::PMC::ptr() },
+            rng: unsafe { &*device::RNG::ptr() },
+            syscon: Syscon::from(syscon),
+        }
     }
 }
 
-#[export_name = "main"]
-fn main() -> ! {
-    let syscon = SYSCON.get_task_id();
-    let syscon = Syscon::from(syscon);
+impl BlockRngCore for Lpc55Core {
+    type Item = u32;
+    type Results = [u32; 1];
 
-    syscon.enable_clock(Peripheral::Rng);
+    fn generate(&mut self, results: &mut Self::Results) {
+        results[0] = self.rng.random_number.read().bits();
+    }
+}
 
-    let rng = unsafe { &*device::RNG::ptr() };
-    let pmc = unsafe { &*device::PMC::ptr() };
+struct Lpc55Rng(BlockRng<Lpc55Core>);
 
-    let mut buffer = [0u32; 1];
+impl Lpc55Rng {
+    fn new() -> Self {
+        Lpc55Rng {
+            0: BlockRng::new(Lpc55Core::new()),
```

Comment:
Unusual to use this style of initialization for a tuple struct.

Reply:
Fixed. There's a clippy lint for this but it's only a few months old and hasn't trickled down yet: https://github.com/rust-lang/rust-clippy/pull/8170